### ISSUE TITLE
Ho to handle JobCancellationException in kotlin ?

### ISSUE THREAD

**Issue by fvigotti** [resilience4j/resilience4j#688]
Hi,
I'm trying to figure out which is the best way to handle JobCancellationException
because this is often an exception that you don't want to retry because probably the main context has been closed ( ie in actor or nested jobs ) 

the problem is that JobCancellationException is `internal` in kotlin so It's not filtrable with `.retryExceptions(..classes..)`
So I'm ended with a filter which leave all exception except unknown CancellationException

```
.retryOnException { t ->
                    val retryVal = when(t){
                        is CancellationException -> when(t){
                            is TimeoutCancellationException -> true
                            else -> false // do not retry other type of cancellation exception ( IE : JobCancellationException )
                        }
                        else -> true
                    }
                    log.info { "retry filter = $retryVal , for exception : $t  " }
                    retryVal
                }
```

I find myself adding this on almost every resilience4j retry object that is used inside coroutines.. so I wonder if someone have a better design for this or maybe if something more generic & beautiful could be implemented in `.executeSuspendFunction` function

**Comment by RobWin** [comment]
You can the following:

```
RetryConfig retryConfig = RetryConfig.custom()
                .retryExceptions(TimeoutCancellationException.class, SomeOtherException).build();

RetryRegistry retryRegistry = RetryRegistry.of(retryConfig)
```

Every instance created from this registry has the same new default configuration.

Is that what you need?

**Comment by RobWin** [comment]
Sry, I misunderstood your issue. Let me think again.

**Comment by RobWin** [comment]
Can you decorate your call and catch and rethrow another custom RuntimeException which can be filtered by Retry.

**Comment by RobWin** [comment]
Did it help you?

**Comment by RobWin** [comment]
I'm fine, If we modify the existing functions that they are not retried when the coroutine was cancelled.

**Comment by RobWin** [comment]
Can you try to record the super-class CancellationException instead, which is a part of the public API.

**Comment by RobWin** [comment]
@ZzAve Is that an issue which you also would like to handle?

**Comment by RobWin** [comment]
/CC @luispollo 
What is your opinion. since you faced the same problems

**Comment by ZzAve** [comment]
I think it it favourable for resilience not to interfere with `CancellationExceptions` unless explicitly requested. Recently I changed the behaviour for the timeLimiter to actually wrap the the `TimeoutCancellationException` in a `java.util.concurrent.TimeoutException`, with a Timelimiter you're explicitly requesting for `TimeoutExceptions`. 

Now for a `Retry`, or a `CircuitBreaker` for that matter you are not interested in Cancellation, as that is typically triggered outside of the `Retry` context. I'm inclined to say that the `Retry` should record a Cancellation as a failure and re-throw the exception without retrying.

To achieves @fvigotti 's goal of retrying on Timeouts, I'd say nest a `TimeLimiter` inside your `Retry` from which you set the Timeout.

**Comment by luispollo** [comment]
@RobWin Thanks for looping me in. It actually turned out that that was not the cause of our particular issue, although it led me to find the resilience4j kotlin extension, which we've since been using, so it's been a useful  investigation nevertheless. FWIW, to @fvigotti's point above, I would agree that providing a convenience function for users who want to retry specifically on coroutine timeouts sounds like it would be useful. Perhaps something like a `retryOnCoroutineTimeouts()` builder method declared in the kotlin extension lib only?

**Comment by piyushmor** [comment]
should be solved by #1428
